[Messenger] Add AMQP adapter

sroze opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | ø
| License       | MIT

- [x] Depends on the Messenger component #24411
- [x] Add tests once we are all happy about the structure

---

In order to give a great DX for simple needs such as sending messages through an AMQP broker such as RabbitMq, we should ship an AMQP adapter for the Messenger component within Symfony Core. It should be as simple as this proposal. We don't need to handle more specific use-cases nor brokers as other adapters such as the [enqueue adapter](https://github.com/sroze/enqueue-bridge) can also be used. 

Configuring the adapter is as simple as the following configuration:
```yaml
# config/packages/messenger_adapters.yaml
framework:
    messenger:
        adapter: "%env(MESSENGER_DSN)%"
```

With the given `.env` for example:
```
MESSENGER_DSN=amqp://guest:guest@localhost:5672/%2f/messages
```

Keep in mind that after having configured the adapter, developers have to route their messages to the given adapter.

```yaml
# config/packages/messenger_routes.yaml
framework:
    messenger:
        routing:
producer).
            'App\Message\Command\CreateNumber': messenger.default_sender
```

---

Additionally, multiple adapters can be created and messages routed to these ones.

```yaml
# config/packages/messenger_routes.yaml
framework:
    messenger:
        adapters:
            commands: "amqp://guest:guest@localhost:5672/%2f/commands"
            maintenance: "amqp://guest:guest@localhost:5672/%2f/maintenance"
        routing:
producer).
            'App\Message\Command\CreateNumber': messenger.commands_sender
            'App\Message\Command\MaintenanceSpecificCommand': messenger.maintenance_sender
```

sroze committed:
[Messenger] Add AMQP adapter

javiereguiluz commented on src/Symfony/Bridge/PhpAmqp/AmqpReceiver.php:
```diff
@@ -0,0 +1,58 @@
+use Symfony\Component\Messenger\Transport\Serialization\DecoderInterface;
+
+/**
+ * Symfony Message receiver to get messages from AMQP brokers using PHP's AMQP extension.
```
`Symfony Message` -> `Symfony Messenger` ?

javiereguiluz commented on src/Symfony/Bridge/PhpAmqp/AmqpSender.php:
```diff
@@ -0,0 +1,45 @@
+use Symfony\Component\Messenger\Transport\Serialization\EncoderInterface;
+
+/**
+ * Symfony Message sender to send messages to AMQP brokers using PHP's AMQP extension.
```
Here too: `Message` -> `Messenger`

javiereguiluz commented on src/Symfony/Bridge/PhpAmqp/AmqpSender.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $encodedMessage = $this->messageEncoder->encode($message);
+
+        $this->connection->publish(
```
This should probably go in one line.

docteurklein replied:
I'm not sure it's conceptually correct to publish to a **connection**.
I would expect to see `exchange->publish()` here (same goes for the queue).

I would have inverted the dependency, by exposing a `new Queue(new Connection)`, not the contrary.

Maybe you wrapped all the API in a `Connection` for ease of use ?

javiereguiluz commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+ */
+class Connection
+{
+    private $amqpConnectionCredentials;
```
Rename this to `$amqpCredentials` to be consistent with the other vars? `$amqpChannel`, `$amqpExchange`, etc.

sroze replied:
Or maybe `$connectionCredentials` ? All the variables prefixed by `amqp` (except this one) actually contains objects coming from PHP's AMQP library.

javiereguiluz commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+     */
+    private $amqpQueue;
+
+    public function __construct(array $amqpConnectionCredentials, string $exchangeName, string $queueName, bool $debug = false)
```
Given how we use classes for everything in this component (and it's great!) what if we replace this PHP array to store credentials by a proper `AMQPCredentials` class?

sroze replied:
As we are directly passing it to the `\AMQPConnection` object, I believe that it's better not to enforce any constraint on what can be in these "credentials". IMHO there is no point of maintaining an object mapping all the options available for the `\AMQPConnection` constructor coming from PHP's AMQP library.

nicolas-grekas commented:
Shouldn't this be *inside* the messenger component?

javiereguiluz commented on src/Symfony/Bridge/PhpAmqp/composer.json:
```diff
@@ -0,0 +1,35 @@
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "Samuel Rozé",
```
If your surname is spelled `Rozé` ... you need to update the PHPdoc of all classes where you spell it as `Roze` 😄

sroze replied:
Yeah... ASCII destroyed my surname ages ago. I gave up on the `é` so will revert to `e` 😛

javiereguiluz commented on src/Symfony/Bundle/PhpAmqpBundle/PhpAmqpBundle.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class PhpAmqpBundle extends Bundle
```
Here we spell `AMQP` as `Amqp` ... which I think it's the official policy of Symfony code syntax. So, in all other classes, we should replace `AMQP` by `Amqp` -> e.g. `AMQPEnvelope` -> `AmqpEnvelope`

sroze replied:
Unfortunately the `\AMQP*` classes such as `AMQPEnvelope` are coming from PHP's AMQP library 🤷‍♂️

theofidry replied:
You can use aliased imports ;)

sroze commented:
> Shouldn't this be inside the Messenger component?

That's a good question. I took the outside approach but it's debatable as the "inside" approach is basically the Cache component's approach.

Cons of having them inside are:
1. You don't want to download an AMQP client when you get the component to use it as a message bus, or when you use another adapter.
2. It will be harder to draw the line of what's "inside" or "outside" regarding the adapters created by the community.

The pros of having them inside:
1. As I foresee many other adapters (for example an "HTTP" adapter - ApiPlatform or not - in order to send/receive messages via HTTP), if we decide that this one should be outside the component, it means more maintenance "cost" for us to add more "official" ones (i.e. creating packages, recipes, bundles, etc...)

With years of having many adapters within the Cache component, what is your PoV? Was that a good choice?

nicolas-grekas commented:
> You don't want to download an AMQP client

We shouldn't care for extra files. We're talking about 500 lines here, LICENSE included.
We just need to make ext-amqp a suggestion instead.

sroze commented:
@nicolas-grekas you are just quoting a small part of my answer here. I understand your answer to my question about Cache Component's choice is "yes, definitely". 

Therefore, let's say we go with the adapter bundled inside, how would you have a similar DX than this one that we can have with another bundle?
1. `composer req amqp`
2. Change `AMQP_DSN` environment variable (if needed)
3. Configure the routing of your message in `config/packages/framework.yaml`

nicolas-grekas commented:
@sroze we'll just ship with a default config based on e.g. PDO. Or no default config if none are available.

sroze commented:
> we'll just ship with a default config based on e.g. PDO

What do you mean? Where would you see this configuration to be tweaked by the developer? Could you give me a bullet-list example compared to what I described above as I don't get how it could be simple for the developer?

nicolas-grekas commented:
> Where would you see this configuration to be tweaked by the developer

in a recipe

theofidry commented:
@nicolas-grekas but if you do `composer req amqp` and you don't have the extension installed it won't be able to pick it up if those things are handled by the recipe whereas if it's outside you can put the extension as a requirement upfront.

xabbuh commented:
@theofidry There can be a pack that also defines the dependeny on the amqp extension and we can then provide a recipe for the pack.

theofidry commented:
But shouldn’t that bridge be also usable from outside of Symfony?

On Thu 22 Mar 2018 at 11:34, Christian Flothmann <notifications@github.com>
wrote:

> @theofidry <https://github.com/theofidry> There can be a pack that also
> defines the dependeny on the amqp extension and we can then provide a
> recipe for the pack.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/symfony/symfony/pull/26632#issuecomment-375273016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE76gfvSvIYfsAOSfGbI2pQ5FTx2wcKXks5tg4w6gaJpZM4S2sxP>
> .
>

sroze commented:
> But shouldn’t that bridge be also usable from outside of Symfony?

Well, there is no point of this adapter/bridge without the Messenger component anyway, isn't it?

Nyholm commented:
> Shouldn't this be inside the Messenger component?

I would say no. 

If it is inside the Messanger component, then why shouldn't we also add a [php-amqplib/php-amqplib](https://github.com/php-amqplib/php-amqplib) adapter? And a php-enqueue adapter? And an adapter for Kafka, Redis etc etc etc. Where do you draw the line?

One could argue that we should just have the **one**, but I think that the AMQP adapter is just a implementation detail (not a core thing) of the component.

makasim commented:
There are three AMQP implementations out there php-amqplib, bunny, amqp-ext and amqp-interop that provides a set of interop interfaces. It would be better to use amqp interop and do not hard code dependency to the extension. Let developers to choose what they want to use

nicolas-grekas commented:
This needs to be in the Messenger component IMHO. We don't ship components that require third-party code, because then we cannot enforce our policies for BC/FC, deprecations, etc.

> Where do you draw the line?

Exactly where our policies need to be enforced so that we can provide the Symfony guarantees.

Nyholm commented:
> We don't ship components that require third-party code

That is the thing. The messanger component does not **need** any AMQP library/extension. That is just an extra feature. 

---

@makasim 👍

nicolas-grekas commented:
@Nyholm ??? yes it *needs* a store, as soon as you need async.

sroze commented:
> Where do you draw the line?

@Nyholm That question is valid regardless of having adapters inside or outside the component: which ones becomes a `symfony/` package and which ones are living somewhere else... I agree with @nicolas-grekas here, it's a must-have in terms of DX for most of the users. And does not cost anything (except few KBs) for others :)

> Let developers to choose what they want to use

@makasim that's exactly what we are doing by exposing all these extension points (i.e. the Sender/Receiver and Encoder/Decoder interfaces). That's exactly why I've built [the Enqueue adapter](https://github.com/sroze/enqueue-bridge). The point here is what should be shipped directly within Symfony Core.

Nyholm commented:
> yes it needs a store, as soon as you need async.

Yes, but it is not obvious that you need async. 

I would still like to see that the adapters is not part of the component package. But if we really feel we want to ship some: lets do php-enqueue and ext-amqp as our "adapters shipped with the component"

dunglas commented:
I'm for storing adapters directly in the component.
It's easier to discover for the end user, and it's consistent with what is done for the Cache component and PropertyInfo component (we provide an adapter for phpDocumentator's ReflectionParser, and it's inside the component directly).

dunglas commented on src/Symfony/Bridge/PhpAmqp/AmqpReceiver.php:
```diff
@@ -0,0 +1,58 @@
+    public function receive(): iterable
+    {
+        while (true) {
+            if (null === ($message = $this->connection->waitAndGet())) {
```
Useless parenthesis.

Nyholm commented:
But what about all external dependencies that different adapters use? We would still need ask the user to download third party code. 

```php
throw new \Exception('To use the XXX adapter you need to download XXX/library');
```

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+        $this->debug = $debug;
+    }
+
+    public static function fromDsn(string $dsn, bool $debug = false)
```
`: self`

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+
+    public static function fromDsn(string $dsn, bool $debug = false)
+    {
+        if (false === ($parsedUrl = parse_url($dsn))) {
```
Useless parenthesis

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+            throw new \InvalidArgumentException(sprintf('The given AMQP DSN "%s" is invalid.', $dsn));
+        }
+
+        $pathParts = explode(trim($parsedUrl['path'] ?? '', '/'), '/');
```
What about `$pathParts = isset($parsedUrl['path'])) ? explode(trim($parsedUrl['path'], '/'), '/') : [];`?

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        $this->exchange()->publish($body, null, AMQP_NOPARAM, array(
+            'headers' => $headers,
```
Can be inlined

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+        $message = null;
+
+        try {
+            $this->queue()->consume(function (\AMQPEnvelope $envelope) use (&$message) {
```
`function (\AMQPEnvelope $envelope) use (&$message): bool`

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+                return false;
+            });
+        } catch (\AMQPQueueException $e) {
+            if (404 == $e->getCode()) {
```
`===`

dunglas commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+
+    private function channel(): \AMQPChannel
+    {
+        if (null === $this->amqpChannel) {
```
What about a guard clause instead to reduce the complexity?

```php
if (null !== $this->amqpChannel) {
    return $this->amqpChannel;
}

$connection = new \AMQPConnection($this->amqpConnectionCredentials);
if (false === $connection->connect()) {
    throw new \AMQPException('Could not connect to the AMQP server. Please verify the provided DSN.');
}

return $this->amqpChannel = new \AMQPChannel($connection);
```

I would do the same for all other methods in this file.

dunglas commented on src/Symfony/Bundle/PhpAmqpBundle/DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getConfigTreeBuilder()
```
`: TreeBuilder`?

dunglas commented on src/Symfony/Bundle/PhpAmqpBundle/DependencyInjection/PhpAmqpExtension.php:
```diff
@@ -0,0 +1,52 @@
+            $container->getParameter('kernel.debug'),
+        )))->setFactory(array(Connection::class, 'fromDsn'));
+
+        $container->setDefinitions(array(
```
I suggest to call `$container->register` several times instead for readability.

joelwurtz commented:
The main problem with the approach of storing adapter in the component is for versioning purpose, let's imagine there is a new version of the amqp extension with a different API. When you update this adapter it will become a problem, sure you can add a conflict version, but you may have other libs depending on the old extension (and not updated). 

Also people would not be able to profit for new features / bug fix unless they update their extension (although it's an implementation detail)

However having an adapter outside this component you would just need to create a new one with the new API and maintain the old one in parallelel (or create some deprecation policy to allow people upgrading to the new one)

dunglas commented:
Versionning isn’t an issue. We can create bridges for both versions of the dep (we did that for Reflection DocBlock in Prophecy for instance).

Regarding extra packages, we’ll add adapters dependencies in the suggest section. And we can always create metapackages such as message-amp for DX of suggesting isn’t enough.

sroze commented:
I agree that the best option is actually to package the AMQP adapter within the Messenger component. This will simplify the maintenance and ease the discoverability of the adapter. In order to integrate it properly with Symfony it, therefore, requires works on the `messenger` section of the FrameworkBundle, like the following:
```yaml
framework:
    messenger:
        adapters:
            default: "amqp://guest:guest@localhost:5672/%2f/messages"
```

@fabpot to prevent conflicts, it seems like the best idea is to wait for the Messenger PR to be merged within `master` :)

docteurklein commented on src/Symfony/Bridge/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,192 @@
+        if (null === $this->amqpExchange) {
+            $this->amqpExchange = new \AMQPExchange($this->channel());
+            $this->amqpExchange->setName($this->exchangeName);
+            $this->amqpExchange->setType(AMQP_EX_TYPE_FANOUT);
```
should the exchange type be configurable?

sroze commented:
Updated to have the adapter within the component. It also exposes another extension point (`messenger.adapter_factory` tag) that allows other adapter factories to be plugged-in very easily and allow users to simply use a DSN-based configuration.

sroze commented:
I've updated the description with the according usage example.

sroze commented:
@nicolas-grekas @dunglas @javiereguiluz @docteurklein @theofidry ready for another round of review on the structure :)

theofidry commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Exception/RejectMessageException.php:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+interface RejectMessageException
```
shouldn't this implement `Throwable`?

sroze replied:
Good idea 👍

makasim commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,207 @@
+        $message = null;
+
+        try {
+            $this->queue()->consume(function (\AMQPEnvelope $envelope) use (&$message): bool {
```
The callback might get a message from another queue (if several subscribed in one process).

makasim replied:
The amqp-ext is really fast though its consume method has one disadvantage it does not react on system signal, so if you have a big (non zero ) read timeout and getting signals they are ignored. bunny and php-amqplib throws an exception in this case.

makasim replied:
@sroze Have you tried to run consume method several times. It might create new subscription on rabbitmq side. 

You can check that by comparing consumer tags or by looking at queue consumer in Rabbit Managment UI.

sroze replied:
Thank you for your review @makasim. 

> The callback might get a message from another queue (if several subscribed in one process).

We don't allow this within the current implementation and wouldn't suggest we do for the long term within this AMQP bridge anyway. I guess this means we don't have to take this into account here.

> The amqp-ext is really fast though its consume method has one disadvantage it does not react on system signal

Might it be different with its version 1.9.x? I've added an integration test to ensure the signals are properly listened to.

sroze replied:
My point was that because we don't allow (yet?) to subscribe from the different queues in the same process, this should be OK.

Though, if we do, what would be your recommendation? In your experience, what's the best way to handle such situation?

makasim replied:
> My point was that because we don't allow (yet?) to subscribe from the different queues in the same process, this should be OK.

You cannot guaranty that, and a developer might take the AMQP channel and subscribe themselves or configure messenger the way that makes it possible. 

Is there a way to change the queue used? 

Once it happens you are in trouble. The behavior is really tricky and hard to reproduce, happens from time to time. I've spent a couple of months to fix it.

lyrixx replied:
I don't really like this design either.

1. It can work with only one queue at a time
2. Can not react (nicely) with signals
3. The PHP process can not does something else between 2 AMQP messages
4. It's blocking

sroze replied:
> It can work with only one queue at a time

Per running process, yes. 

> The PHP process can not does something else between 2 AMQP messages

What would it do? :)

> Can not react (nicely) with signals

It does stop to consume when receiving the signal.

Though, any alternative suggestion? 😃

makasim replied:
> It does stop to consume when receiving the signal.

It is not true for non zero timeout.

sroze replied:
Fair enough. So should we force a zero-timeout or do you guys see another solution?

makasim replied:
Do you have some kind of extensions or plugins that should be executed from time to time disregarding there is a message or not?

lyrixx replied:
> What would it do? :)

For exemple, In a specific app I worked on, the worker looks at the filesystem to check if something changed.

> Though, any alternative suggestion?

We use to use the `get` method instead of the `consume` method.

> Do you have some kind of extensions or plugins that should be executed from time to time disregarding there is a message or not?

1. Monitoring: send a ping to consul (for exemple) to said "I'm alive"
2. Check the memory of the current script. If > 100Mb (for exemple), commit suicide

sroze replied:
> We use to use the get method instead of the consume method.

In this context, I'm not sure to understand what would this change? Because as a developer, you wouldn't have access to this call either way, because this `waitAndGet` method is called inside the `AmqpReceiver`. If I understand your point correctly, this would mean that you'd like the `AmqpReceiver` to sometimes yield `null`s so you can do something else in between by creating a decorator, right? 🤔

lyrixx replied:
That's an idea. 

my point is: a consumer (cli process) will be very coupled with the underlying fetcher. 
It's not a daemon where we can do anything.

sroze replied:
After reflexion, that's a good point here. What if we simply `yield` `null`s [in the `AmqpReceiver` instead of `continue`ing](https://github.com/symfony/symfony/pull/26632/files#diff-da5657a5035db784d7c45b1b168d65fdR45) ? That use in combination with a timeout (so it actually yield nulls) and you're sorted out by simply creating a `ReceiverInterface` decorator, right?

makasim commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,207 @@
+        if (null === $this->amqpChannel) {
+            $connection = new \AMQPConnection($this->connectionCredentials);
+
+            if (false === $connection->connect()) {
```
It is better to use persisted connection. Or leave an option for a developer

sroze replied:
Great. I've added that feature and a test for it 👍

chalasr commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Exception/RejectMessageException.php:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+interface RejectMessageException extends \Throwable
```
Throwable cannot be extended https://3v4l.org/E7eQ6 , also it misses the interface suffix

sroze replied:
That's the point: this interface should only be implemented by classes that are extending `\Exception`: https://3v4l.org/belYf.

Agreed with the `Interface` suffix though.

chalasr replied:
Right!

sroze commented on .travis.yml:
```diff
@@ -12,6 +12,7 @@ addons:
         - language-pack-fr-base
         - ldap-utils
         - slapd
+        - librabbitmq-dev
```
Yep, we use the `\AMQP*` classes within the tests.

chalasr replied:
Then they need the `@requires extension amqp` annotation :)

sroze replied:
Good point. Added 👍

sroze commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1445,6 +1445,22 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
         $container->getDefinition('messenger.asynchronous.routing.sender_locator')->replaceArgument(1, $messageToSenderIdsMapping);
+
+        foreach ($config['adapters'] as $name => $adapter) {
+            $container->setDefinition($adapterId = 'messenger.'.$name.'_adapter', (new Definition(\Symfony\Component\Messenger\Adapter\Factory\AdapterInterface::class, array(
```
`AdapterInterface` is already `use`d because of the Cache component. I'll alias it.

chalasr commented on src/Symfony/Component/Messenger/Tests/Adapter/PhpAmqp/AmqpReceiverTest.php:
```diff
@@ -0,0 +1,110 @@
+            new SerializerComponent\Serializer(array(new ObjectNormalizer()), array('json' => new JsonEncoder()))
+        );
+
+        $envelope = $this->createMock(\AMQPEnvelope::class);
```
I believe we use `getMockBuilder(class)->getMock()` for broader phpunit compat, dunno if it's encore required, but at least we don't have any `createMock()` occurrence in the codebase

sroze replied:
Changed all of these.

thePanz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -966,6 +966,22 @@ private function addMessengerSection(ArrayNodeDefinition $rootNode)
+                            return is_array($v) && isset($v['adapter']);
+                        })
+                        ->then(function ($v) {
+                            $v['adapters'] = array('default' => $v['adapter']);
```
Shouldn't it use the short syntax for arrays here?

javiereguiluz replied:
Yes! However, to keep consistency with all the existing PHP code, we decided to keep using the long `array()` syntax instead of `[...]` even for the new code added to the master branch (which requires PHP 7.1). It's shown here: https://symfony.com/doc/current/contributing/code/standards.html

thePanz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1445,6 +1445,22 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+        foreach ($config['adapters'] as $name => $adapter) {
+            $container->setDefinition($adapterId = 'messenger.'.$name.'_adapter', (new Definition(\Symfony\Component\Messenger\Adapter\Factory\AdapterInterface::class, array(
+                $adapter['dsn'],
+            )))->setFactory(array(new Reference('message.adapter_factory'), 'create'))->setPublic(true));
```
Shouldn't it use the short syntax for arrays here?

chalasr replied:
No, symfony doesn't use short array syntax yet for maintenance reasons.

sroze commented:
@dunglas @chalasr can you have another look please?

chalasr commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,210 @@
+    public function ack(\AMQPEnvelope $message)
+    {
+        return $this->queue()->ack($message->getDeliveryTag());
+    }
```
we group methods by visibility, public before private (same for properties/consts)

sroze replied:
Changed :+1:

chalasr reviewed (approved):
1 minor comment left, rebase needed and 👍

sroze commented:
Rebased and updated based on comments. Ready 👍

yceruto commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -73,5 +73,18 @@
         </service>
+
+        <!-- Adapters -->
+        <service id="message.adapter_factory" class="Symfony\Component\Messenger\Adapter\Factory\ChainAdapterFactory">
```
``messenger.adapter_factory``?

yceruto commented on src/Symfony/Component/Messenger/Adapter/Factory/ChainAdapterFactory.php:
```diff
@@ -0,0 +1,50 @@
+            }
+        }
+
+        throw new \InvalidArgumentException(sprintf('The given DSN "%s" is invalid', $dsn));
```
missing dots at the end?

yceruto commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpReceiver.php:
```diff
@@ -0,0 +1,66 @@
+            }
+
+            try {
+                $shouldStop = self::STOP === (yield $this->messageDecoder->decode(array(
```
Is this statement correct? it seems that `(yield ...)` will be `null` always https://3v4l.org/51jhv.

Could we do this `yield $shouldStop = $this->messageDecoder->decode(...)` and check it below? ... or throw a new `StopReceiveException` instead?

yceruto replied:
Oh NVM, I see it https://github.com/symfony/symfony/pull/26632/files#diff-b1417962dd38ee12497279be6f3f1308R53 I guess we need to do that (`$g->send('stop')`) to stop the iterator here.

lyrixx replied:
So the only way to stop the iteration is via the iterator ? It could be convenient to also expose a method `stop()`. We could use this method in a listener (or something like that) that watch the memory for exemple. Or to have a nice control of signals

sroze replied:
Why do we have this `->send('stop')` ? The problem is that if we use a `break` to leave the `foreach` around the generator, it will not execute what is after the `yield` in this file, and therefore will not acknowledge the message. 

Having a `stop` method on the receiver is a more explicit alternative but it can't be "forwarded" by decorating generators without knowing the receiver (not that it's a massive problem as I believe all of the decorations will be decorators on the `ReceiverInterface`). 

That said, do you have any other idea on how to tackle this problem than adding a `close` method?

lyrixx replied:
> That said, do you have any other idea on how to tackle this problem than adding a close method?

Unfortunatly, I don't have a solution with the existing design :/ In my previous PR I managed to fix it by decoupling the "Loop" and the "Consumer"

sroze replied:
I think that `ReceiverInterface::stop` is a good idea. We wouldn't propagate the stop of the consumer back via the Generator but simply by calling `$receiver->stop()`, which would, in turn, change the `$shouldStop` boolean within the consumer. Simpler.

makasim commented:
There are BC breaks coming with [2.0 version](https://github.com/pdezwart/php-amqp/issues/257) of the extension. You'd have to break it too. 

This wouldn't be the case if you adopt [AMQP Interop](https://github.com/queue-interop/queue-interop#amqp-interop). As it, just interfaces\contract and any compatible adapter could be plugged in. 

I confident in it and can release stable release anytime If it something that stop you from using it.

sroze commented:
Thanks again for your review @makasim. To ensure the version of the php-amqp extension, we can use an `amqp` meta-package, but this adds a bit of complexity. Is there any way to detect its version number at runtime? I can't see any constant defined 🤔 

No doubt we will promote the enqueue adapter (which we agreed will be in the enqueue organisation 🎉), though (so far) we'd like to have a very minimalist AMQP adapter within the Core.

chalasr commented:
@sroze can't we add a conflict rule for `ext-amqp:<1.9.3`?

sroze commented:
Yep, very good idea. I guess `<1.9.3,>=2.0.0` will prevent us from the BCs of the extension 2.0 as well.

Tobion commented:
Could you give an example how to set options for AMQP like `x-dead-letter-exchange` and `x-message-ttl`? Those are often needed. So there should be an easy way to set those.

sroze commented:
@chalasr added the conflict. 
@Tobion good point. Added it in https://github.com/symfony/symfony/pull/26632/commits/587c741c892f056d9ab7e7fe6caa83408382421b 👍

lyrixx commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpFactory.php:
```diff
@@ -0,0 +1,35 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Adapter\PhpAmqp;
```
Why did you use this namespace `PhpAmqp` ? 
I guess it's because it's for the  PHP AMQP PECL
I said that because it could be confused with https://github.com/php-amqplib/php-amqplib

I would named it `Amqp`

sroze replied:
Fair enough, this could be confused, but `Amqp` is as vague. What about `AmqpExt`?

lyrixx commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpReceiver.php:
```diff
@@ -0,0 +1,66 @@
+            try {
+                $shouldStop = self::STOP === (yield $this->messageDecoder->decode(array(
+                    'body' => $message->getBody(),
+                    'headers' => $message->getHeaders(),
```
Why don't you passe the full messages ? I could be needed to build a custom retry-strategy

sroze replied:
Because I would be AMQP-specific. The contract of the encoder/decoder is a plain PHP array, nothing more: this can support different type of messages. Though, I'm sure we could add much more of the message's properties to it if necessary (and I suggest we do this when adding the retry-strategy) in the next iteration :)

lyrixx commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpReceiver.php:
```diff
@@ -0,0 +1,66 @@
+
+                throw $e;
+            } catch (\Throwable $e) {
+                $this->connection->nack($message);
```
I would requeue the message, because if you don't, the message will be lost

sroze replied:
Updated 👍

lyrixx commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,218 @@
+    private $connectionCredentials;
+    private $exchangeConfiguration;
+    private $queueConfiguration;
+    private $debug;
```
Can you sort this list of variables with the same order of the constructor parameters?

lyrixx commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,218 @@
+        return $this->queue()->reject($message->getDeliveryTag());
+    }
+
+    public function nack(\AMQPEnvelope $message)
```
You need to expose the `\AmqpQueue::nack` $amqpParam

sroze replied:
Good point, done 👍

lyrixx commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,218 @@
+        $this->exchange()->declareExchange();
+
+        $this->queue()->declareQueue();
+        $this->queue()->bind($this->exchange()->getName());
```
This works because the default exchange type is Fan Out. If one use Direct is will not work anymore. You may add a routing key option to configure it

sroze replied:
Good point, added the routing key as a possible option on the queue. `queue[routing_key]=...` in the DSN 👍

lyrixx reviewed (commented):
I left few comments

sroze commented on src/Symfony/Component/Messenger/Adapter/Factory/AdapterFactoryInterface.php:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+interface AdapterFactoryInterface
```
The thing is: for the developer experience, I believe that we need to have a very common way of creating the adapters: via a simple DSN, and that's the point of this (thin) adapter factory. That way, it's super easy for other developers to create new adapters.

If we move the `supports` down to the `AdapterInterface`, then we need to move the `$dsn` argument to both the `receiver` and `sender` methods and sort of force a double-parsing of it on the developer's perspective. I don't mean that it's impossible but less elegant I'd say.

lyrixx replied:
The parsing could (should ?) be done in the FrameworkExtension ?

sroze replied:
If it's only for "native" adapters, yes. But this can't be done for adapters coming from 3rd party libraries, and that's the entire point: we want 3rd party component to plug-in and have people using a simple DSN such as `enqueue+sqs://sqs-address` for example. While this wouldn't be in the core obviously.

lyrixx replied:
To me, it is still another layer of indirection. But I'm not strongly against this. Let's see what other think.

sroze replied:
You are right. It's "an extension point" we don't need now as just the adapter factory is enough. Removed in https://github.com/symfony/symfony/pull/26632/commits/d6c046aed17efca71cc112fa4dd88783cfe74119.

sroze commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpAdapter.php:
```diff
@@ -0,0 +1,50 @@
+
+    public function receiver(): ReceiverInterface
+    {
+        return new AmqpReceiver($this->decoder, $this->connection);
```
This is actually registered as a service, and therefore executed once by the DIC :)

lyrixx replied:
The `AmqpAdapter` is registered in the DIC, not the `AmqpReceiver` isn't ?

sroze replied:
It is: check [this](https://github.com/symfony/symfony/pull/26632/files#diff-0e793081ceb720201745c982a568903fR1470) out. The adapter service is the factory of the receiver and sender services.

lyrixx replied:
Sorry, I guess I missed something :( The `AmqpAdapter`, the sender and the receiver are indeed registred in the DIC. This **this** very `AmqpAdapter` does not use the the sender or the receiver that are in the DIC.

sroze replied:
This very `AmqpAdapter` is used to create the sender and receiver services. Therefore `$container->get('messenger.[name]_sender')` will get this very `AmqpSender` object.

Though, you are right that `$this->get('messenger.[name]_adapter')->receiver()` will return a new instance, and this might be a cause of issues. I guess I'll add a `if (null === $this->receiver) ...` in the adapter so it won't re-create it.

sroze commented:
Thank you very much @lyrixx for the review. I've updated a few things and added comments to the rest. Can you check them out? :)

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1459,6 +1460,23 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+            $container->setDefinition($adapterId = 'messenger.'.$name.'_adapter', (new Definition(MessengerAdapterInterface::class, array(
+                $adapter['dsn'],
+                $adapter['options'],
+            )))->setFactory(array(new Reference('messenger.adapter_factory'), 'create'))->setPublic(true));
```
There is no correlation between an extension point and the fact it's private or public.

sroze replied:
It's trickier for people to use the service if it isn't public 😄

lyrixx commented:
@sroze Nice iteration. Thanks.

sroze commented on .travis.yml:
```diff
@@ -12,11 +12,13 @@ addons:
     global:
         - MIN_PHP=7.1.3
         - SYMFONY_PROCESS_PHP_TEST_BINARY=~/.phpenv/shims/php
+        - MESSENGER_AMQP_DSN=amqp://localhost/%2f/messages
```
Though, in [RabbitMq's defaults](https://www.rabbitmq.com/access-control.html#default-state), the default virtual host is named `/`. That's why I used it and I believe as it matches the default, this is the one to keep.

soyuka commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,218 @@
+        $this->amqpFactory = $amqpFactory ?: new AmqpFactory();
+    }
+
+    public static function fromDsn(string $dsn, array $options = array(), bool $debug = false, AmqpFactory $amqpFactory = null): self
```
What about a Trait or some reusable method for the parsing stuff?

sroze replied:
I'd use the same point used in the previous comment: [we don't know yet so let's not expose too much](https://github.com/symfony/symfony/pull/26632#discussion_r178805235). Or said another way: what is public (i.e. to be able to be shared/re-use) is supersensible to BC-break and adds up to our BC responsibility. Adding such exposure for an hypothetic use case is too risky IMHO.

sroze commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,218 @@
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class Connection
```
> every implementation will have: ack, reject, nack, fromDsn methods

We don't know yet. Tbh, I'm convinced the other way around: most won't. But either way, I suggest we continue to improve this later (i.e. potentially adding an interface) when we see the real use case. (also, this `Connection` object is only interesting for the `Amqp` adapter).

sroze commented:
@lyrixx @soyuka @makasim thank you so much for your tests and reviews that pushes that adapter and the transport's design to their limits. This (hopefully last) change should resolve the last blockers.

Following the discovery of [this `yield`/`try`/`catch` weirdness](https://gist.github.com/sroze/2ed22483b5fd35081b4fa31f02fd1a24) (which [isn't a bug](https://bugs.php.net/bug.php?id=76181)) I decided to question this choice that I loved of using an `iterable` as a result of the `ReceiverInterface`. The only way to manage errors is via sending the exceptions back to the generator (via `$generator->throw($exception)`): this adds a lot of complexity (i.e. `if instanceof Generator` checks in the `catch` of `ReceiverInterface` decorators) and allows an easy way to break this error management by forgetting to forward the error down or by simply not using generators anymore. Therefore, I changed the interface to take a message handler instead of returning that generator. It looks good (`$receiver->receive($handler)` is literally correct) and prevents a weird management of exceptions.

Last but not least, in order to prevent the issues mentioned by Max and tackle in Greg's PR, I moved from using the `consume` method to the `get` method. We now have to manually handle the signals but first, we can be sure it's properly handled (see [the test](https://github.com/symfony/symfony/pull/26632/files#diff-31d54404cd29fcd612f175ed67b18216R67)) and it removes this very specific constraint on the AMQP library version.

sroze commented on src/Symfony/Component/Messenger/Worker.php:
```diff
@@ -35,22 +35,22 @@ public function __construct(ReceiverInterface $receiver, MessageBusInterface $bu
+        $this->receiver->receive(function($message) {
+            if (null === $message) {
+                return;
+            }
```
Because the receiver's `receive` callback can receive a `null` message in case of timeout. This is to tackle [the point exposed by @lyrixx](https://github.com/symfony/symfony/pull/26632#discussion_r178811238): that way, you can add extra capabilities (like sending pings to the monitoring, etc...) at timely intervals even if no messages are handled. You basically put a `read_timeout` on the connection and decorate the receiver to do whatever you want (with or without messages coming in).

Nyholm commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd:
```diff
@@ -368,6 +369,20 @@
+        <xsd:sequence>
+            <xsd:element name="option" type="messenger_adapter_option" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+
```
Remove blank line

Nyholm commented on src/Symfony/Component/Messenger/Adapter/Factory/AdapterInterface.php:
```diff
@@ -0,0 +1,25 @@
+use Symfony\Component\Messenger\Transport\SenderInterface;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
```
I would really like a line or two on this interface to explain what it does.

Nyholm reviewed (approved):
I've reviewed the changes and I like it. I've not tested it and played around with the implementation yet.

sroze commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php:
```diff
@@ -0,0 +1,219 @@
+        }
+
+        try {
+            if (false !== $message = $this->queue()->get()) {
```
Obviously, there are pros and cons to both methods. Your previous reviews suggested we shouldn't use the `consume` method as it wasn't reliable and had some ["strange behaviour"](https://github.com/symfony/symfony/pull/26632#discussion_r178153477). Let's try to get as close to the fact as possible. Here is a comparison table of the two approaches:

| | `consume` method  | `get` methods |
| -| ------------ | ------------- |
| **Pros**  | <ul><li>Is blocking (low CPU impact for the loop)</li></ul>  | <ul><li>Easy signal handling capabilities</li><li>Will be relatively easy to support `get`ing from multiple queues</li><li>Allow us to adds `ticks` while waiting for the messages</li></ul>
| **Cons**  | <ul><li>Does not handle signals well</li><li>Can't easily handle multiple queues (as blocking)</li><li>Has ["strange behaviour"](https://github.com/symfony/symfony/pull/26632#discussion_r178153477) for ext-amqp < 1.9.3 (which is the last one)</li><li>Messages [might be mixed up](https://github.com/symfony/symfony/pull/26632#discussion_r176930611) between queues</li></ul>  | <ul><li>Higher CPU footprint because of the loop</li><li>It not visible in RabbitMq's UI</li><li>Does not support QoS</li></ul>

makasim replied:
There is no any strangeness in `basic.conume`, it comes from the way it was used and there is a solution for it a buffer.

I am all for basic consume properly applied .

sroze replied:
That said, I suggest we go for the `get` method option which seems to be the one that will provide the greater stability. Later, if we some use-cases/issues and believe this was not the correct choice, we can add an option to use the `consume` method instead easily. That way we move forward and stay open to changes :)

ogizanagi commented on src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpReceiver.php:
```diff
@@ -0,0 +1,80 @@
+                throw $e;
+            }
+
+            if (function_exists('pcntl_signal_dispatch')) {
```
Not in a `finally`? Not needed?

sroze replied:
That's a good shot, this could be in the `finally`. I don't know if it's really needed but doesn't cost :)

sroze commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -970,12 +970,21 @@ private function addMessengerSection(ArrayNodeDefinition $rootNode)
                     ->{!class_exists(FullStack::class) && class_exists(MessageBusInterface::class) ? 'canBeDisabled' : 'canBeEnabled'}()
+                    ->validate()
+                        ->ifTrue(function ($v) { return isset($v['adapter']) && isset($v['adapters']); })
+                        ->thenInvalid('"adapter" and "adapters" cannot be used together.')
```
True, I removed it. It appears that DX-wise, it does not change anything actually thanks to the recipe (i.e. `framework.messenger.adapters.default` configured by default - maybe commented). And if we put `framework.messenger.adapter` in the recipe, it's less obvious they can use multiple adapters.

sroze commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1470,6 +1470,24 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
         }
+
+        foreach ($config['adapters'] as $name => $adapter) {
+            $container->setDefinition('messenger.'.$name.'_sender', (new Definition(SenderInterface::class))->setFactory(array(
```
Happy to change but... What do you mean by "usually"? :)

ogizanagi replied:
I mean that's AFAIK what most service factories will choose as format for their ids (see security listeners concrete instances or user_checkers named after firewalls, workflows, Monolog handlers, ...) but also what we do anyway to gather similar services under an identifiable "namespaces".

sroze replied:
Fair enough. Changed.

sroze commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php:
```diff
@@ -543,6 +544,33 @@ public function testMessengerValidationDisabled()
 
+    public function testMessengerAdapter()
+    {
+        $container = $this->createContainerFromFile('messenger_adapter');
```
Good catch, will handle that in another PR 😉

sroze replied:
You fixed it in #26816, thank you :)

sroze commented on src/Symfony/Component/Messenger/Adapter/Factory/AdapterFactoryInterface.php:
```diff
@@ -0,0 +1,29 @@
+
+    public function createSender(string $dsn, array $options): SenderInterface;
+
+    public function supports(string $dsn, array $options): bool;
```
Tbh, you can skip the `AdapterFactoryInterface` if you don't want the developers to use the DSN to create your senders/receivers. Registering and tagging the sender/receiver still works. Also, you could create a `NullSender` or `NullAdapter`.

sroze commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/yml/messenger_adapter.yml:
```diff
@@ -0,0 +1,8 @@
+            customised:
+                dsn: 'amqp://localhost/%2f/messages?exchange_name=exchange_name'
+                options:
+                    queue_name: Queue
```
Added 👍

davidbarratt commented:
Any chance on this getting merged soon? Thanks! :)

davidbarratt commented on src/Symfony/Component/Messenger/Adapter/AmqpExt/AmqpFactory.php:
```diff
@@ -0,0 +1,35 @@
+
+    public function createQueue(\AMQPChannel $channel): \AMQPQueue
+    {
+        return new \AMQPQueue($channel);
```
I think these `\AMQP` classes are only in the `amqp` pecl extension? if so, it should be a requirement in `composer.json`
https://getcomposer.org/doc/04-schema.md#package-links
something like
```
"ext-amqp": "*"
```
unless there's an alternative?

sroze replied:
That's correct. We can't add the requirement in this `composer.json` file as it's not a requirement for the entire component, but for its AMQP adapter only. Though you are right in pointing that we don't handle properly the case when the extension is not installed: we can add a nice exception message when trying to create the adapter while the extension does not exist.

davidbarratt replied:
good to know! I'm not going to hold up the train for that, but probably a good idea to add at some point.
